src/blockdata/transaction.rs

```diff
@@ -220,12 +221,166 @@ impl Default for TxIn {
         TxIn {
             previous_output: OutPoint::default(),
             script_sig: Script::new(),
-            sequence: u32::max_value(),
+            sequence: Sequence::MAX,
             witness: Witness::default(),
         }
     }
 }
 
+/// Bitcoin transaction input sequencing number.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Debug, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
+pub struct Sequence(pub u32);
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+/// An error in creating relative lock-times.
+pub enum RelativeLockTimeError {
+    /// The input was too large
+    InputTooLarge(u32)
+}
+
+impl Sequence {
+    /// The maximum allowable sequence number
+    pub const MAX: Self = Sequence(MAX_SEQUENCE);
+    /// The lowest sequence number that does not opt-in for replace-by-fee.
+    pub const RBF_SIGNAL: Self = Sequence(0xFFFFFFFE);
```

Comment:
If this does not opt-in, then why is it called `RBF_SIGNAL`?

I would assume using this would signal (opt-in) for RBF.

Reply:
Perhaps `RBF_THRESHOLD`?

```suggestion
    /// A transaction is considered to have opted in to allowing replacement of itself if any of its
    /// inputs have a `Sequence` number less than this threshold (Explicit Signalling [BIP-125]).
    ///
    /// [BIP-125]: <https://github.com/bitcoin/bips/blob/master/bip-0125.mediawiki]>
    pub const RBF_THRESHOLD: Self = Sequence(0xFFFFFFFE);
```

Reply:
The main use for MAX - 1 is to enable nLockTime usage without activating RBF.

So maybe `ENABLE_LOCKTIME_NO_RBF` is better? (though it doesn't feel right to me either)

Reply:
`NO_RBF_MASK`? (ie. if `NO_RBF_MASK & self != NO_RBF_MASK` then it is RBF)

Again... this also sounds awkward.

Reply:
Perhaps have two const with the same value because the following code is going to read funny
```rust
    if n < Sequence::ENABLE_LOCKTIME_NO_RBF {
        // not at all clear what this scope means
    }
```

Reply:
`ENABLE_LOCKTIME_NO_RBF` reads nicely in struct declarations but the other usecase of checking that sequence is below this value will read a little strangely
```rust
if n < Sequenece::ENABLE_LOCKTIME_NO_RBF {
    // What does this scope mean?
}
```

Reply:
I like `ENABLE_LOCKTIME_NO_RBF`, but it will have to either be `ENABLE_ABSOLUTE_LOCKTIME_NO_RBF` or have absolute documented to avoid confusion around relative and absolute lock-times.

I also do echo @tcharding's point about using the constant when checking. We do provide a function to do checks already (`is_rbf()`) but for any reason users want to check themselves or have some other logic using the constant, the name doesn't provide much meaning.

Reply:
I am not a user of the library currently, but in BitcoinJS the point in code where developers "decide if we want to use RBF" is when creating the TxIn structs, and it's usually just a tertiary op between MAX and MAX - 2.

If Rust is similar, and we assume most people won't be using OP_CSV OP_CLTV, then it might make sense to say:

1. `Default::default()` or `Sequence::MAX` is default.
2. `Sequence::ENABLE_ABSOLUTE_LOCKTIME_NO_RBF` for people who want to be able to use nLockTime in an old-fashioned way (no OP code) without enabling RBF
3. `Sequence::ENABLE_RBF_NO_TIMELOCK_SCRIPTS` (MAX - 2) for people who "just want to enable RBF"
4. Documentation mentions that Everything else is RBF enabled.

The documentation for the `Sequence` struct should be fleshed out to explain RBF. Then all of the consts should make mention of what they enable/disable in regard to a. RBF b. nLockTime usage c. OP_CSV/OP_CLTV.

Just my two cents.

If you would like I can write up some docs and make a PR to your branch.

Reply:
Yeah, public constants should be used for constructing, not for comparing. Methods are better for comparing. But we use the constants for comparing internally and would like to have them readable too. Maybe duplicate private constants with better names?

Reply:
Public constants from @junderw for constructing and something along the lines of `Sequence::RBF_THRESHOLD` for internal use maybe?

Also agree with making the docs on the struct and constants better. I'll document what the struct is used for (lock-times and rbf) and mark the constants with what they enable/disable.

Reply:
How is the following?
```rust
/// Bitcoin transaction input sequencing number.
/// 
/// The sequence field is used for:
/// - Indicating whether [BIP-65] absolute lock-times are enabled.
/// - Indicating and encoding [BIP-68] relative lock-times.
/// - Indicating whether a transcation opts-in to [BIP-125] replace-by-fee.
/// 
/// [BIP-65]: <https://github.com/bitcoin/bips/blob/master/bip-0065.mediawiki>
/// [BIP-68]: <https://github.com/bitcoin/bips/blob/master/bip-0068.mediawiki>
/// [BIP-125]: <https://github.com/bitcoin/bips/blob/master/bip-0125.mediawiki>
#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Debug, Hash)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
pub struct Sequence(pub u32);

// ...

/// The maximum allowable sequence number.
/// 
/// Having this sequence number has no effect on the transaction.
pub const MAX: Self = Sequence(MAX_SEQUENCE);
/// The lowest sequence number that does not opt-in for replace-by-fee.
/// 
/// A transaction is considered to have opted in to replacement of itself
/// if any of it's inputs have a `Sequence` number less than this value
/// (Explicit Signalling [BIP-125]).
/// 
/// [BIP-125]: <https://github.com/bitcoin/bips/blob/master/bip-0125.mediawiki]>
pub const RBF_THRESHOLD: Self = Sequence(0xFFFFFFFE);
/// The sequence number that enables OP_CLTV usage but disables replace-by-fee and relative lock-time.
pub const ENABLE_LOCKTIME_NO_RBF: Self = Sequence::RBF_THRESHOLD;
/// The sequence number that enables replace-by-fee but disables use of lock-times.
pub const ENABLE_RBF_NO_LOCKTIME: Self = Sequence(0xFFFFFFFD);
```

I've kept `RBF_THRESHOLD` public if users want to use it for their own comparisons.

Reply:
I like the top-most doc!

>  Having this sequence number has no effect on the transaction.

Not true, it disables absolute lock time.

> I've kept RBF_THRESHOLD public if users want to use it for their own comparisons.

I worry it'll be confusing having two constants for same value. Users should just use methods instead of comparing.

Reply:
Threshold means "upper bound"

It is "upper bound + 1" though.

Even for internal use threshold doesn't make much sense to me.

Reply:
@Kixunil 
> Not true, it disables absolute lock time.

I wrote it meaning lock time cannot affect the transaction but will change it as it is confusing.

Reply:
How about `Sequence::MIN_NO_RBF`?

```rust
pub fn is_rbf(&self) -> bool {
    *self < Sequence::MIN_NO_RBF
}
```

Reply:
`MIN_NO_RBF` sounds good!